Warn if the title is a duplicate

stacey-gammon opened the pull request:
Fixes #10073

<img width="516" alt="screen shot 2017-02-13 at 2 12 45 pm" src="https://cloud.githubusercontent.com/assets/16563603/22898769/81b68d1e-f1f6-11e6-83d9-f22888526e90.png">

Stacey-Gammon commented:
jenkins test this

LeeDr commented:
Does this PR implement the same functionality for Saved Searches and Visualizations?  I didn't do a thorough code review but noticed some dashboard files changed but not others that jumped out as Discover or Visualization changes.

I'm not sure the message is clear enough.  It tells them that there already is a dashboard with this title.  But it could be more clear about what will happen when I click Save dashboard.  It's going to create another dashboard with the same name.

Maybe something like this?
`A dashboard with the title 'New Dashboard' already exists.  Would you like to make a duplicate with the same name?`

But I need to think through the scenarios where I think this is going to appear.

1. one case, I'm editing an object that has been saved and I click Save, leave the existing name, don't check the Save as new ___ checkbox, and click Save.
This just updates the existing object with any changes.  No need to show the dialog.

2. I'm editing an object that has been saved and I click Save, leave the existing name, and I check the Save as new ___ checkbox, and click Save.  That should show your new dialog.

3. I create a new object and the first time I go to save it I enter the same name as an existing object (it could be the default name).  That should show your new dialog.  So I guess this is the same case as 2 above.

Stacey-Gammon commented:
@LeeDr Updated the error message.  I was already handling #1 and #3 but good catch on #2.  Implemented that and added a test for it.

Yes the change is implemented across the board for all objects, as the important changes are in saved_object.js.  I only added tests for the dashboard.

Stacey-Gammon commented:
Hmmm, my query is not correct as it returns matches on substrings.  Trying to figure out how to create an exact match query...

Stacey-Gammon commented:
jenkins test this

 (looks like the build was aborted)

Stacey-Gammon commented:
Hmm, latest build failure is related to saving, but not sure it's due to this PR:

<img width="1172" alt="screen shot 2017-02-16 at 10 18 08 am" src="https://cloud.githubusercontent.com/assets/16563603/23027086/48aab710-f431-11e6-9387-7054154c9ed7.png">
```
15:12:57.770: --- tryForTime failure: [POST http://localhost:4444/wd/hub/session/251abda956b6c9bad5a51ef85b848bc7/element / {"using":"link text","value":"Visualization DataTable"}] no such element: Unable to locate element: {"method":"link text","selector":"Visualization DataTable"}
  (Session info: chrome=55.0.2883.87)
  (Driver info: chromedriver=2.24.417424 (c5c5ea873213ee72e3d0929b47482681555340c3),platform=Linux 3.13.0-74-generic x86_64)
15:13:08.328: --- tryForTime failed again with the same message  ...
15:13:18.888: --- tryForTime failed again with the same message  ...
15:13:29.451: --- tryForTime failed again with the same message  ...
15:13:30.017: Taking screenshot "/var/lib/jenkins/workspace/elastic+kibana+pull-request+multijob-selenium/test/screenshots/failure/failure_1487258010016_data table_should be able to save and load.png"
>> FAIL: chrome on any platform - kibana - visualize app - visualize app - data table - should be able to save and load (49214ms)
Error: tryForTime timeout: [POST http://localhost:4444/wd/hub/session/251abda956b6c9bad5a51ef85b848bc7/element / {"using":"link text","value":"Visualization DataTable"}] no such element: Unable to locate element: {"method":"link text","selector":"Visualization DataTable"}
  (Session info: chrome=55.0.2883.87)
  (Driver info: chromedriver=2.24.417424 (c5c5ea873213ee72e3d0929b47482681555340c3),platform=Linux 3.13.0-74-generic x86_64)
    at runRequest (/var/lib/jenkins/workspace/elastic+kibana+pull-request+multijob-selenium/node_modules/leadfoot/Session.js:88:40)
    at /var/lib/jenkins/workspace/elastic+kibana+pull-request+multijob-selenium/node_modules/leadfoot/Session.js:109:39
```

I'll rebase

cjcenizal commented on src/ui/public/courier/saved_object/saved_object.js:
```diff
@@ -337,7 +380,11 @@ export default function SavedObjectFactory(esAdmin, kbnIndex, Promise, Private,
+            err.message === OVERWRITE_REJECTED ||
+            err.message === SAVE_DUPLICATE_REJECTED)) {
+            return;
+          }
```
Can we break this up to make it a little more readable?

```javascript
if (err) {
  if (err.message === OVERWRITE_REJECTED) {
    return;
  }

  if (err.message === SAVE_DUPLICATE_REJECTED) {
    return;
  }
}
```

Stacey-Gammon replied:
ehhh, I don't love suggestion, I'd look at that and be like, why not just combine them into an ||?

how about this new variation as a compromise?

cjcenizal commented on src/ui/public/courier/saved_object/ui/saved_object_save_as_checkbox.html:
```diff
@@ -3,7 +3,7 @@
   </div>
   <label>
-    <input type="checkbox" ng-model="savedObject.copyOnSave" ng-checked="savedObject.copyOnSave">
+    <input type="checkbox" data-test-subj="saveAsNewCheckbox" ng-model="savedObject.copyOnSave" ng-checked="savedObject.copyOnSave">
```
Can we put these attributes on multiple lines?

```html
<input
  type="checkbox"
  data-test-subj="saveAsNewCheckbox"
  ng-model="savedObject.copyOnSave"
  ng-checked="savedObject.copyOnSave"
>
```

cjcenizal commented on test/functional/apps/dashboard/_dashboard_save.js:
```diff
@@ -0,0 +1,79 @@
+    expect(countOfDashboards).to.equal(2);
+  });
+
+  bdd.it('Does not warn when saving a duplicate title that remains unchanged for an existing dashboard', async function() {
```
I had a hard time parsing this description... how does this sound to you: "Does not warn when you save an existing dashboard with the title it already has, and that title is a duplicate"? Does that describe this test accurately?

cjcenizal commented on test/functional/apps/dashboard/_dashboard_save.js:
```diff
@@ -0,0 +1,79 @@
+    expect(isConfirmOpen).to.equal(false);
+  });
+
+  bdd.it('Warns when saving a duplicate title that remains unchanged when Save as New Dashboard is checked', async function() {
```
And here, how about: "Warns you when you Save as New Dashboard, and the title is a duplicate"?

cjcenizal commented on test/functional/index.js:
```diff
@@ -29,13 +29,13 @@ define(function (require) {
+      // 'intern/dojo/node!./apps/console',
       'intern/dojo/node!./apps/dashboard',
-      'intern/dojo/node!./status_page'
+    //  'intern/dojo/node!./status_page'
```
Haha I do this all the time! 😄

Stacey-Gammon replied:
This time it was actually intentional :)  Trying to reduce the time for a build because I was having a very tough time figuring out what the issue was (turns out it was a delay between saving a new dashboard and reloading the url to point to the newly saved dashboard).

cjcenizal commented on test/support/page_objects/dashboard_page.js:
```diff
@@ -276,8 +319,19 @@ export default class DashboardPage {
     PageObjects.common.debug('Storing time with dashboard: ' + on);
+    const saveAsNewCheckbox = await PageObjects.common.findTestSubject('saveAsNewCheckbox');
+    const checked = await saveAsNewCheckbox.getProperty('checked');
+    if (checked === true && on === false ||
```
Can this be made terser?
```javascript
if (checked !== isAlreadyChecked) {
  PageObjects.common.debug('Flipping save as new checkbox');
  await saveAsNewCheckbox.click();
}
```

Stacey-Gammon replied:
Doh! Good catch, much clearer. :)

cjcenizal commented on test/support/page_objects/dashboard_page.js:
```diff
@@ -276,8 +319,19 @@ export default class DashboardPage {
   }
 
-  async storeTimeWithDashboard(on) {
+  async saveAsNewCheckbox(on) {
```
Can we rename this to `setSaveAsNewCheckBox(checked)` to make it clear what the action is that we're taking?

cjcenizal commented on test/support/page_objects/dashboard_page.js:
```diff
@@ -276,8 +319,19 @@ export default class DashboardPage {
+  async saveAsNewCheckbox(on) {
     PageObjects.common.debug('Storing time with dashboard: ' + on);
+    const saveAsNewCheckbox = await PageObjects.common.findTestSubject('saveAsNewCheckbox');
+    const checked = await saveAsNewCheckbox.getProperty('checked');
```
Can we rename this to `isAlreadyChecked`?

cjcenizal commented on src/ui/public/courier/saved_object/saved_object.js:
```diff
@@ -18,6 +18,28 @@ import MappingSetupProvider from 'ui/utils/mapping_setup';
+ * @param error {Error} the error
+ * @return {boolean}
+ */
+function getShouldSuppressExpectedError(error) {
```
How about renaming this to `isErrorNonFatal`? This way the name of function helps us assign meaning to the error without specifying what we should do about it. If for some reason we decide to return the error instead of suppressing it, this function will still be useful and won't need to be renamed.

Stacey-Gammon replied:
I think that's a little confusing too because it's not an error at all, it's a user response.  The whole flow is a bit awkward, and it'd be better if we just stopped using the confirmModalPromise to begin with, but I don't want to do that refactor now.

 I have no better suggestion, so sgtm for now.

Stacey-Gammon commented:
@cjcenizal  or @LeeDr, any more comments on this PR?  I think I addressed everything.

LeeDr commented:
I just tried a case with saving Visualizations where it didn't work, probably because of the characters I used?
1. I saved a search as `----- #####`
2. I saved it again and checked the box ` Save as a new visualization`
3. I should have gotten the message that it already exists, but I didn't. 

Obviously this is an odd case, and I only tried this one visualization name so far.  So I don't know exactly what cases work and which don't.  Are we comparing analyzed names which would remove those chars?

I'll keep testing a bit more.

LeeDr commented:
Also if I name an object `/////` I can create more of them with no warnings.

But combinations of special characters and letter always seems to work correctly.
`- test - name` gives a warning if I try to make a copy.
and `test - name` doesn't incorrectly give a warning matching `- test - name`.

And I also tested that the duplication check is only within that object type.  So I can create a saved search, visualization, and dashboard all with the name `test1` with no warnings (as I would expect).

If the special character cases aren't easy to solve, I'd still give a LGTM.

Stacey-Gammon commented:
Unfortunately it's not easy to fix @LeeDr.  If I have a visualization name `///` and I run this query in devtools:
```
GET _search
{
  "query": {
    "bool": {
     "must": { "match_phrase": {  "title": "///" } }
    }
  }
}
```
Nothing is returned, presumably because it strips special characters.  

If this issue were resolved: https://github.com/elastic/kibana/issues/10348 then we could get around this shortcoming.

For the short term, I think we should acknowledge it as a limitation, but move forward.
